### DISCUSSION TITLE
HomeAssistant tasmota integration and Berry driver

### DISCUSSION THREAD

**Discussion by msetina** [arendst/Tasmota#20062] [Q&A]
Using a Berry driver to produce sensor values does not produce sensor values as it does for known directo HW drivers like DS18B20. Is there a way to provide value type and unit for a sensor produced by Berry driver?

**Comment by barbudor** [comment]
If your Berry driver starts from autoexec.be, it should work
Tasmota Discovery message is built from driver's response. If you load you driver manually it's too late as the Discovery message as been sent.

**Reply by msetina** [reply]
It is not that the sensors are not discovered. I have the autoexec.be and the driver loads. HA integration finds all the sensors/values. The problem is that the values are not presented to the HA system the same way as normal other drivers. So the presentation in HA is limited to say the least.

**Reply by msetina** [reply]
It looks like I am issing a way to send some sort of "metadata" of the sensor to the HA integration to fully use the HA presentation capabilities.

**Comment by sfromis** [comment]
In that case, you are not doing it right. In addition to having the `json_append` callback, you should also make sure to have your driver initialized during the boot process, so that the **first** sensor cycle includes your driver. When you do that, the driver JSON output is included with other sensor drivers, also for the sensor discovery payload.

**Reply by msetina** [reply]
I have all that and the values/sensors are seen in the HA. The problem is that for example temperature value is not presented in HA as temperature and the unit is not known to the HA. The driver itself and the framework tasmota provides is great. Now I want to be able to present them similar to other drivers that tasmota provides and I fall short. Is there a standard way for a driver to present the "metadata" of sensors to tasmota integration?

**Reply by sfromis** [reply]
No way of metadata, outside of what follows from sensor names like "Humidity". Unfortunately, the setup has the "[Tasmota Integration](https://github.com/emontnemery/hatasmota)" on the HA side include a data structure with sensor names and units to match.

**Reply by barbudor** [reply]
Not knowing how you are presenting your data into your SENSOR message, it's difficult to help
But normally, if you have in your SENSOR message something like `..., "MySensor": { "Temperature": 33.3 }, ..." then HA should properly interpret that as a Temperature
If not ask there as it is not a Tasmota issue

**Reply by msetina** [reply]
Thanks. Will go there. What I can see now is that all the values my berry driver presents get shown the same way as Id of DS18B20. Independant of if the value serialized in JSON is a string or a decimal value and the name as I have a value named Temperature. I am guessing that in the discovery process data of the driver  sensors are sent in MQTT and I do not utilize that.

**Reply by msetina** [reply]
As an example this is how my addition to JSON looks like:
`"HeatContainer":{"00000689D934":{"Energy flow in cell":579.718,"Ratio":1.29348,"Height":2.3,"Temperature":45.4,"Target energy":761.915,"Level":9.6,"Energy to inlet":985.521,"Energy to target":-223.606}}`

**Reply by sfromis** [reply]
Are you sure that the Tasmota integration actually supports the extra object level in JSON you have there, when you clearly deviate from the normal Tasmota sensor JSON format. And spaces in field names may be quite "risky" too for having it supported.

**Reply by msetina** [reply]
I am not sure what a standard format is. Now I found some example formats in hatasmota sensor.py file, but in the same file I read that units are read from discovery which I do not remember that it can be manipulated in Berry driver. I assume there is also some naming convention I failed to find.

**Reply by sfromis** [reply]
Standard format is like other sensors. One single JSON level below the sensor name, and the sensor object is at top JSON level, like this:
`"VEML7700":{"Illuminance":0,"WhiteContent":0}`

The way you are (at least should be) manipulating the discovery payload is not direct, but indirectly by producing the regular JSON format, without additional actions. The discovery payload for sensors is the regular sensor JSON wrapped inside an extra object level:
topic `tasmota/discovery/xxxxxxxxxxxx/sensors`
payload `{"sn":{"Time":"2023-11-19T13:04:03.365+01:00","DS18B20":{"Id":"xxxxxxxxxxxx","Temperature":11.69}, ....`
Do not expect this to work when creating an alternate format with the extra object level.

**Reply by barbudor** [reply]
Temperature unit will automatically be appended by Tasmota as soon as there is a "Temperature" field in the JSON

**Reply by sfromis** [reply]
This implies that you should ideally make sure to respect SetOption8 and supply the temperature value in C/F depending on this setting. Supporting TempOffset and TempRes too would make it work like the standard drivers of Tasmota. Here's a driver doing this:
https://github.com/TomsTek/tasmota-berry-TMP117-driver/blob/main/tmp117.be

**Reply by msetina** [reply]
OK. So we can have a sensor with an identifing name and a set of values that need to have names that will be mapped to units inside tasmota (or HA integration) to send/use as discovery payload. 
Let me redo my side to confirm.

**Reply by msetina** [reply]
Thank you. Got most of them with nested example. A have no idea how to make Height, Level and Percentage/Ratio register properly.
`"HeatCell2": {
        "Id": "0000067001EE",
        "Temperature": {
            "Now": 28.1
        },
        "Level": 2.5,
        "Energy": {
            "ToInlet": 0.360073,
            "Total": 0.895682,
            "ToTarget": 0.535609
        },
        "Height": 2.5,
        "Percentage": 40.201
    },
    "HeatContainer": {
        "Percentage": 76.2745,
        "Temperature": {
            "Output": 40,
            "Target": 40
        },
        "Height": 11.7,
        "Energy": {
            "ToInlet": 3.19727,
            "ToTarget": 0.994523,
            "ForOutput": 4.19179,
            "Total": 4.19179,
            "FlowInCell": 0.809985
        }
    },`

**Reply by msetina** [reply]
The HA integration does not recognize Height, Level and Percentage/Ratio. WIll go to integration guys to see what can be done.

**Reply by sfromis** [reply]
With `Temperature` the "Tasmota way" of handing the unit (C/F) is to have `"TempUnit":"C"` in the JSON, which Tasmota should automatically provide for you **if you follow the standard format**.

**Reply by msetina** [reply]
The full JSON is very long so I posted just an excerpt and at the end the TempUnit is provided. I am questioning if other Units can be provided. OTH there are other things HA integration for tasmota sets from code...

**Reply by sfromis** [reply]
What is available there is only:
```
PressureUnit
SpeedUnit
TempUnit
```
This is **only** the fields where Tasmota has options for user selection of what unit to use, in general the units are assumed to be "self-evident".
